src/qibo/optimizers.py

```diff
@@ -36,32 +38,56 @@ def cma(loss, initial_parameters, options=None):
         options (dict): Dictionary with options accepted by the ``cma``.
             optimizer. The user can use ``cma.CMAOptions()`` to view the
             available options.
+        args (tuple): optional arguments for the loss function.
     """
     import cma
-    r = cma.fmin2(loss, initial_parameters, 1.7, options=options)
+    r = cma.fmin2(loss, initial_parameters, 1.7, options=options, args=args)
     return r[1].result.fbest, r[1].result.xbest
 
 
-def newtonian(loss, initial_parameters, method='Powell', options=None):
+def newtonian(loss, initial_parameters, method='Powell', options=None, processes=None, args=()):
     """Newtonian optimization approaches based on ``scipy.optimize.minimize``.
 
     For more details check the `scipy documentation <https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html>`_.
 
+    .. note::
+        When using the method ``parallel_L-BFGS-B`` the ``processes`` option controls the
+        number of processes used by the parallel L-BFGS-B algorithm through the ``multiprocessing`` library.
+        By default ``processes=None``, in this case the total number of logical cores are used.
+        Make sure to select the appropriate number of processes for your computer specification,
+        taking in consideration memory and physical cores. In order to obtain optimal results
+        you can control the number of threads used by each process with the ``qibo.set_threads`` method.
+        For example, for small-medium size circuits you may benefit from single thread per process, thus set
+        ``qibo.set_threads(1)`` before running the optimization.
+
     Args:
         loss (callable): Loss as a function of variational parameters to be
             optimized.
         initial_parameters (np.ndarray): Initial guess for the variational
             parameters.
-        method (str): Name of method supported by ``scipy.optimize.minimize``.
+        method (str): Name of method supported by ``scipy.optimize.minimize`` and ``'parallel_L-BFGS-B'`` for
+            a parallel version of L-BFGS-B algorithm.
         options (dict): Dictionary with options accepted by
             ``scipy.optimize.minimize``.
+        processes (int): number of processes when using the parallel BFGS method.
+        args (tuple): optional arguments for the loss function.
     """
-    from scipy.optimize import minimize
-    m = minimize(loss, initial_parameters, method=method, options=options)
+    if method == 'parallel_L-BFGS-B':
+        from qibo.config import raise_error, get_device, get_threads, log
+        if processes is None and get_threads() != 1: # pragma: no cover
```

Comment:
I would change this check to

```suggestion
        if ((processes is not None and processes * get_threads() > psutil.cpu_count())
            or processes is None and get_threads() != 1): # pragma: no cover
```
to also capture cases where both `processes` and threads are set by the user but their product is higher than the available number of logical threads. For example someone may do `qibo.set_threads(8)` and call the optimizer with `processes=10` which will most likely lead to poor performance if the machine has less than 80 threads, but it will not be captured by the check as it is now.

Also, although not very important, I would move this check after the GPU check/error. Now if someone uses this with a GPU it will most likely raise both the warning and the GPU RuntimeError, however the warning will not be relevant in this case.

Reply:
Thanks, that's a quite good remark.